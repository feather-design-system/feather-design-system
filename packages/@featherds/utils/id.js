const getSafeId = function (str) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
