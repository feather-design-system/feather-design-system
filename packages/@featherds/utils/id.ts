const getSafeId = function (str: string) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
