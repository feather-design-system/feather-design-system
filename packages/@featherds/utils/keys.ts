const KEYCODES = {
  PAGEUP: 33,
  PAGEDOWN: 34,
  UP: 38,
  DOWN: 40,
  LEFT: 37,
  RIGHT: 39,
  HOME: 36,
  END: 35,
  ENTER: 13,
  SPACE: 32,
  ESCAPE: 27,
  BACKSPACE: 8,
  DELETE: 46,
};

export { KEYCODES };
