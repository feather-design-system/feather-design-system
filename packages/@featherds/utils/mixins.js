var translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (key) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};
export { translateKey };
