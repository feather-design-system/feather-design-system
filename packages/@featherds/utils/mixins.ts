const translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (this: IComponent, key: string) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};

interface IComponent {
  translationKey: string;
  $t: (s: string) => void;
}
export { translateKey };
