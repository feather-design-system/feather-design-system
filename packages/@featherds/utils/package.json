{
  "name": "@featherds/utils",
  "version": "0.12.31",
  "publishConfig": {
    "access": "public"
  },
  "author": "NantHealth",
  "scripts": {
    "compile": "tsc --declaration"
  },
  "keywords": [
    "vue",
    "FeatherUtils"
  ]
}
