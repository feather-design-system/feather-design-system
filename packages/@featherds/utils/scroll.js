var toView = function (el, parent) {
  if (!el || !parent) {
    return;
  }
  var elHeight = el.getBoundingClientRect().height;
  const compStyle = getComputedStyle(el);
  elHeight += parseInt(compStyle["margin-top"], 10);
  elHeight += parseInt(compStyle["margin-bottom"], 10);
  parent.scrollTop =
    el.offsetTop - parent.getBoundingClientRect().height + elHeight;
};

export { toView };
