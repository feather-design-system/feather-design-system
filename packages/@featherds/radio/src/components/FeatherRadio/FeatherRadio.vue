<template>
  <div class="layout-container">
    <div
      class="feather-radio"
      role="radio"
      ref="input"
      :aria-checked="checked ? 'true' : 'false'"
      :aria-disabled="disabled ? 'true' : 'false'"
      :aria-labelledby="labelId"
      :tabindex="tabindex"
      @click="click"
      @blur="blur"
      data-ref-id="feather-radio"
    >
      <div class="radio hover focus">
        <div class="icon" aria-hidden="true">
          <svg height="16" width="16" class="dot">
            <circle cx="8" cy="8" r="5" stroke="black" stroke-width="0" />
          </svg>
        </div>
        <feather-ripple center v-if="!disabled" />
      </div>

      <span class="label" data-ref-id="feather-radio-label" :id="labelId">
        <slot />
      </span>
    </div>
  </div>
</template>
<script>
import { getSafeId } from "@featherds/utils/id";
import { FeatherRipple } from "@featherds/ripple";
import { computed, inject, ref } from "vue";
export default {
  props: {
    value: undefined,
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  setup(props) {
    const first = ref(false);
    const checked = ref(false);
    const labelId = computed(() => getSafeId("radio-label-id"));
    const tabindex = computed(() => {
      if (first.value) {
        return 0;
      }
      return checked.value ? 0 : -1;
    });

    const input = ref(null);
    const focus = () => {
      input.value.focus();
    };

    //register
    const register = inject("register");
    const blur = inject("blur");
    const select = inject("select");

    const vm = {
      first,
      focus,
      disabled: props.disabled,
      value: props.value,
      checked,
    };
    register(vm);
    const click = () => {
      select(vm);
    };

    return {
      labelId,
      tabindex,
      first,
      blur,
      click,
      input,
      checked,
    };
  },
  components: {
    FeatherRipple,
  },
};
</script>

<style lang="scss" scoped>
@import "~@featherds/styles/mixins/typography";
@import "~@featherds/styles/themes/variables";
@import "~@featherds/styles/themes/utils";

.feather-radio {
  &:focus {
    outline: 0;
  }
  @include state-on-surface();
}
.layout-container {
  display: inline-block;
  line-height: 0;
}

.feather-radio {
  cursor: pointer;
  display: inline-flex;
  align-items: center;
  margin-left: -8px;
  .radio {
    height: 36px;
    width: 36px;
    display: inline-block;
    padding: 8px;
    border-radius: 100%;
  }
  .icon {
    height: 20px;
    width: 20px;
    border: 2px solid var($secondary-text-on-surface);
    background-color: transparent;
    border-radius: 100%;
    position: relative;
  }
  .dot {
    position: absolute;
    height: 16px;
    width: 16px;
    fill: var($primary);
    border-radius: 100%;
    opacity: 0;
    transform: scale(0);
  }
  .label {
    @include body-small();
    color: var($primary-text-on-surface);
    min-width: 24px;
    padding-right: 12px;
    padding-left: 4px;
  }
}

.feather-radio:hover {
  &[aria-disabled="true"] {
    cursor: default;
  }
}

.feather-radio[aria-checked="true"] {
  .icon {
    border-color: var($primary);
  }
  .dot {
    transition: all 280ms ease-in-out;
    opacity: 1;
    transform: scale(1);
  }
}
.feather-radio[aria-disabled="true"] {
  @include state-disabled();
  .icon {
    border-color: var($disabled-text-on-surface);
  }
  .label {
    color: var($disabled-text-on-surface);
  }
  .dot {
    fill: var($disabled-text-on-surface);
  }
}
</style>
