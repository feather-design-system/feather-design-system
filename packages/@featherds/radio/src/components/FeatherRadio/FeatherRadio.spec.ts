import { mount } from "@vue/test-utils";
import FeatherRadio from "./FeatherRadio.vue";
import axe from "@featherds/utils/test/axe";
import * as id from "@featherds/utils/id";
import { nextTick } from "vue";
import { vi, expect, describe, it } from "vitest";
vi.spyOn(id, "getSafeId").mockImplementation((x) => x);

const getWrapper = function (options: Record<string, unknown> = {}) {
  const props = (options.props as Record<string, unknown>) || {};

  options.props = {
    ...props,
    value: true,
  };
  return mount(FeatherRadio, options);
};
const getSlot = () => ({
  slots: {
    default: "Hello",
  },
});
const mockProvide = () => {
  const register = vi.fn();
  const select = vi.fn();
  const blur = vi.fn();
  return {
    provide: {
      register,
      select,
      blur,
    },
  };
};

describe("FeatherRadio.vue", () => {
  it("should call register", () => {
    const { provide } = mockProvide();
    getWrapper({ ...getSlot(), global: { provide } });
    expect(provide.register).toHaveBeenCalled();
  });
  it("should be focusable when first is set even if not checked", async () => {
    const wrapper = getWrapper({ ...getSlot(), global: { ...mockProvide() } });
    wrapper.vm.vm.first = true;
    await nextTick();
    expect(wrapper.vm.tabindex).toBe(0);
  });
  it("should call blur", () => {
    const { provide } = mockProvide();
    const wrapper = getWrapper({ ...getSlot(), global: { provide } });
    (wrapper.vm as unknown as { blur: () => void }).blur();
    expect(provide.blur).toHaveBeenCalled();
  });

  it("should call select from provide", () => {
    const { provide } = mockProvide();
    const value = 1;
    const wrapper = getWrapper({
      ...getSlot(),
      props: {
        value,
      },
      global: { provide },
    });

    wrapper.vm.click();
    expect(provide.select).toHaveBeenCalled();
  });

  it("should select element on click", async () => {
    const value = 1;
    const { provide } = mockProvide();
    const wrapper = getWrapper({
      ...getSlot(),
      props: {
        value,
      },
      global: { provide },
    });

    await wrapper.find("[role='radio']").trigger("click");
    expect(provide.select).toHaveBeenCalled();
  });
  it("should set aria-checked based on data", async () => {
    const wrapper = getWrapper({
      ...getSlot(),
      global: { ...mockProvide() },
    });
    wrapper.vm.vm.checked = true;
    await nextTick();
    expect(wrapper.element).toMatchSnapshot();
    wrapper.vm.vm.checked = false;
    await nextTick();
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should set aria-disabled based on property", async () => {
    const wrapper = getWrapper({
      ...getSlot(),
      props: {
        disabled: true,
      },
      global: { ...mockProvide() },
    });
    expect(wrapper.element).toMatchSnapshot();
    await wrapper.setProps({
      disabled: false,
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  describe("a11y", () => {
    it("be accessible when unchecked and enabled", async () => {
      const wrapper = getWrapper({
        ...getSlot(),
        global: { ...mockProvide() },
      });
      wrapper.vm.vm.checked = false;
      await nextTick();
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
    it("be accessible when checked and enabled", async () => {
      const wrapper = getWrapper({
        ...getSlot(),
        global: { ...mockProvide() },
      });
      wrapper.vm.vm.checked = true;
      await nextTick();
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
    it("be accessible when checked and disabled", async () => {
      const props = { disabled: true };
      const wrapper = getWrapper({
        ...getSlot(),
        props,
        global: { ...mockProvide() },
      });
      wrapper.vm.vm.checked = true;
      await nextTick();
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
    it("be accessible when unchecked and disabled", async () => {
      const props = { disabled: true };
      const wrapper = getWrapper({
        ...getSlot(),
        props,
        global: { ...mockProvide() },
      });
      wrapper.vm.vm.checked = false;
      await nextTick();
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
  });
});
