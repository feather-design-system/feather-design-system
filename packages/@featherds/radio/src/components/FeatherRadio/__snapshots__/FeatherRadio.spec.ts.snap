// Vitest Snapshot v1

exports[`FeatherRadio.vue > should set aria-checked based on data 1`] = `
<div
  class="layout-container"
>
  <div
    aria-checked="true"
    aria-disabled="false"
    aria-labelledby="radio-label-id"
    class="feather-radio"
    data-ref-id="feather-radio"
    id="feather-radio-button"
    role="radio"
    tabindex="0"
  >
    <div
      class="radio hover focus"
      style="overflow: hidden; position: relative;"
    >
      <div
        aria-hidden="true"
        class="icon"
      >
        <svg
          class="dot"
        >
          <circle
            cx="50%"
            cy="50%"
            r="0.3125rem"
            stroke="black"
            stroke-width="0"
          />
        </svg>
      </div>
      <!---->
    </div>
    <span
      class="label"
      data-ref-id="feather-radio-label"
      id="radio-label-id"
    >
      
      Hello
      
    </span>
  </div>
</div>
`;

exports[`FeatherRadio.vue > should set aria-checked based on data 2`] = `
<div
  class="layout-container"
>
  <div
    aria-checked="false"
    aria-disabled="false"
    aria-labelledby="radio-label-id"
    class="feather-radio"
    data-ref-id="feather-radio"
    id="feather-radio-button"
    role="radio"
    tabindex="-1"
  >
    <div
      class="radio hover focus"
      style="overflow: hidden; position: relative;"
    >
      <div
        aria-hidden="true"
        class="icon"
      >
        <svg
          class="dot"
        >
          <circle
            cx="50%"
            cy="50%"
            r="0.3125rem"
            stroke="black"
            stroke-width="0"
          />
        </svg>
      </div>
      <!---->
    </div>
    <span
      class="label"
      data-ref-id="feather-radio-label"
      id="radio-label-id"
    >
      
      Hello
      
    </span>
  </div>
</div>
`;

exports[`FeatherRadio.vue > should set aria-disabled based on property 1`] = `
<div
  class="layout-container"
>
  <div
    aria-checked="false"
    aria-disabled="true"
    aria-labelledby="radio-label-id"
    class="feather-radio"
    data-ref-id="feather-radio"
    id="feather-radio-button"
    role="radio"
    tabindex="-1"
  >
    <div
      class="radio hover focus"
    >
      <div
        aria-hidden="true"
        class="icon"
      >
        <svg
          class="dot"
        >
          <circle
            cx="50%"
            cy="50%"
            r="0.3125rem"
            stroke="black"
            stroke-width="0"
          />
        </svg>
      </div>
      <!--v-if-->
    </div>
    <span
      class="label"
      data-ref-id="feather-radio-label"
      id="radio-label-id"
    >
      
      Hello
      
    </span>
  </div>
</div>
`;

exports[`FeatherRadio.vue > should set aria-disabled based on property 2`] = `
<div
  class="layout-container"
>
  <div
    aria-checked="false"
    aria-disabled="false"
    aria-labelledby="radio-label-id"
    class="feather-radio"
    data-ref-id="feather-radio"
    id="feather-radio-button"
    role="radio"
    tabindex="-1"
  >
    <div
      class="radio hover focus"
      style="overflow: hidden; position: relative;"
    >
      <div
        aria-hidden="true"
        class="icon"
      >
        <svg
          class="dot"
        >
          <circle
            cx="50%"
            cy="50%"
            r="0.3125rem"
            stroke="black"
            stroke-width="0"
          />
        </svg>
      </div>
      <!---->
    </div>
    <span
      class="label"
      data-ref-id="feather-radio-label"
      id="radio-label-id"
    >
      
      Hello
      
    </span>
  </div>
</div>
`;
